src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,300 @@
+                LeaseRevokeRequest.newBuilder()
+                        .setID(leaseId)
+                        .build();
+        return leaseFutureStub.leaseRevoke(leaseRevokeRequest);
+    }
+
+    /**
+     * keep alive one lease in background
+     * @param leaseId
+     */
+    @Override
+    public void keepAlive(long leaseId) {
+        synchronized (lock){
+            if(keepAlives.containsKey(leaseId)){
+
+            }else{
+                Lease lease = new Lease(leaseId);
+                lease.setNextKeepAlive(System.currentTimeMillis());
+                lease.setDeadLine(System.currentTimeMillis() + firstKeepAliveTimeOut);
+                keepAlives.put(leaseId, lease);
+            }
+        }
+
+        return;
+    }
+
+    /**
+     * keep alive one lease in background
+     * @param leaseId
+     */
+    @Override
+    public ListenableFuture<LeaseKeepAliveResponse> keepAliveOnce(long leaseId) {
+
+        /**
+         * to be completed, I will return a ListenableFuture value in the future
+         */
+        StreamObserver<LeaseKeepAliveRequest> requestObserver =
+                leaseStub.leaseKeepAlive(keepAliveResponseStreamObserver);
+        requestObserver.onNext(newKeepAliveRequest(leaseId));
+        requestObserver.onCompleted();
+
+
+        return null;
+    }
+
+    /**
+     *Scan all the leases and seed keep alive request to etcd server
+     */
+    private void keepAliveExecutor(){
+        synchronized (lock){
+            long now = System.currentTimeMillis();
+            List<Long> toSendIds = new ArrayList<>();
+            for(Lease l : keepAlives.values()){
+                if(now > l.getNextKeepAlive()){
+                    toSendIds.add(l.getLeaseID());
+                }
+            }
+
+            for(Long id: toSendIds){
+                keepAliveRequestStreamObserver.onNext(newKeepAliveRequest(id));
+            }
+        }
+    }
+
+    /**
+     * Scan all the leases, remove the dead leases and notify with LeaseHandler
+     */
+    private void deadLineExecutor(){
+        synchronized (lock){
+            long now = System.currentTimeMillis();
+            List<Long> expireLeases = new ArrayList<>();
+            for(Lease l : keepAlives.values()){
+                if(now > l.getDeadLine()){
+                    expireLeases.add(l.getLeaseID());
+                }
+            }
+
+            for(Long id: expireLeases){
+                removeLease(id);
+                if(leaseHandler!=null){
+                    leaseHandler.onLeaseExpired(id);
+                }
+            }
+        }
+    }
+
+    /**
+     * This method update the deadline and NextKeepAlive time
+     * @param leaseKeepAliveResponse The response receive from etcd server
+     */
+    public void processKeepLiveRespond(LeaseKeepAliveResponse leaseKeepAliveResponse){
+        synchronized (lock){
+            if(keepAlives.containsKey(leaseKeepAliveResponse.getID())){
```

Comment:
If it doesn't contain the key, is it something dangerous? Can we throw an exception? Or at least do some logging?